File: app/models/tool_configs/tool_chain.py
"""
Nesting something inside a tool config — the edge, and nothing else.

A tool config answers one question with one query. A **link** says that something
else's result restricts it: the child runs first, one named column of its result
becomes a list of values, and the parent's query is filtered to those values. That
is a sub-query expressed as two reusable things rather than as one large statement,
which is the point — the child stays callable on its own, keeps its own
description, and can be embedded by more than one parent.

**A child is either another tool config or a Graph Designer graph.** The second is newer
and changes nothing about what the edge *means*: a graph is still a thing that runs first
and produces values. What it adds is that a graph may stop mid-run to ask a person a
question, which no tool config can do — so a chain has a third outcome besides rows and
"nothing matched", and ``tool_chain_graph`` carries it as the question itself.

The relationship is a row and not a JSON list on the parent, unlike
``ChatbotFlow.graph_data``, because three things need to *query* it rather than
read one parent's copy of it: the list page (which tools embed which), the agent
runtime (a parent's agent gains every transitive child), and the delete guard (a
tool that something embeds may not be removed). A JSONB list of child uuids would
make each of those a scan.

**A link is a filter, so losing one silently is the failure mode to design
against.** If a link disappeared while its parent kept running, the parent's
result would quietly widen — the same reason
``app.services.deep_agents.query_executor`` fails a tool loudly rather than
dropping a filter whose column was switched off. Both foreign keys cascade so a
deleted agent or datasource cannot strand rows, and ``tool_chain_service`` refuses
to delete or disable a tool that is embedded, which is where that guarantee
actually lives.

The direction of the edge is parent → child, and the tree is walked depth-first at
run time by :mod:`app.services.tool_configs.tool_chain_graph`, which turns it into
LangGraph nodes. Cycles are refused when a link is saved (see
``tool_chain_service.replace_child_links``); nothing at run time re-checks, because
a cycle that reached the graph would be a bug in that validator, not a state to
recover from.
"""

import uuid as uuid_pkg
from typing import Optional

from sqlalchemy import (
    BigInteger,
    CheckConstraint,
    DateTime,
    ForeignKey,
    Index,
    Integer,
    String,
    UniqueConstraint,
    func,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


# How a child's values reach its parent. Two shapes, and the difference is not a
# detail of binding — it decides how many times the parent runs.
#
#   in_list → the whole list at once, as an expanding IN. The parent runs once.
#             What every link did before this column existed, and still the
#             default: one round trip, one result set, no multiplication.
#
#   each    → one value at a time, bound as a plain scalar. The parent runs once
#             per value and the rows are concatenated. Needed whenever the value
#             is not on the right-hand side of an IN — `dd.id = :x`, or inside a
#             string the database builds, as in
#             `LIKE CONCAT('%s:1:"', :x, '"%')` — because an expanding parameter
#             always renders parenthesised and is a syntax error in both.
#
# Stored as a String rather than a native enum for the same reason `query_mode`
# is: adding a third shape should be a constant and a validator, not a migration
# that rewrites a type every table using it has to be locked for.
BINDING_MODE_IN_LIST = "in_list"
BINDING_MODE_EACH = "each"

BINDING_MODES = (
    (BINDING_MODE_IN_LIST, "Match any of the values"),
    (BINDING_MODE_EACH, "Run once per value"),
)

BINDING_MODE_VALUES = frozenset(value for value, _ in BINDING_MODES)


class ToolConfigLink(Base):
    """
    One thing embedded inside a tool: *the values ``child.child_column`` returns
    restrict ``parent`` at ``parent_reference``.*

    **The child is either another tool config or a Graph Designer graph**, and exactly
    one of ``child_id`` / ``child_graph_id`` is set — enforced by
    ``ck_tool_config_links_one_child``. Two nullable columns rather than one column plus
    a discriminator, because a foreign key is the thing that keeps a child from being
    deleted out from under its parent, and a discriminator would give up both foreign
    keys to save one column.

    The edge means the same thing either way, which is the point: something runs first,
    one named part of its result becomes a list of values, and the parent's query is
    filtered to them. What differs is only what "runs first" involves — one query, or a
    whole drawn graph that may stop to ask a question.
    """

    __tablename__ = "tool_config_links"

    id: Mapped[int] = mapped_column(
        BigInteger,
        primary_key=True,
        autoincrement=True,
    )

    uuid: Mapped[uuid_pkg.UUID] = mapped_column(
        UUID(as_uuid=True),
        default=uuid_pkg.uuid4,
        unique=True,
        index=True,
        nullable=False,
    )

    # The tool that embeds. Its rows are what the agent finally sees.
    parent_id: Mapped[int] = mapped_column(
        BigInteger,
        ForeignKey("tool_configs.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )

    # The tool that runs first and supplies values. NULL when the child is a graph.
    child_id: Mapped[Optional[int]] = mapped_column(
        BigInteger,
        ForeignKey("tool_configs.id", ondelete="CASCADE"),
        nullable=True,
        index=True,
    )

    # The graph that runs first and supplies values. NULL when the child is a tool
    # config. Exactly one of the two is set — see the class docstring and
    # `ck_tool_config_links_one_child`.
    #
    # CASCADE, matching `child_id`: a deleted graph takes its links with it rather than
    # leaving a parent whose filter silently disappeared, which is the failure mode this
    # whole module is designed against. `tool_chain_service` refuses the delete first,
    # where the message can name the parent — the cascade is the backstop for a row
    # removed some other way.
    child_graph_id: Mapped[Optional[int]] = mapped_column(
        BigInteger,
        ForeignKey("tool_graphs.id", ondelete="CASCADE"),
        nullable=True,
        index=True,
    )

    # Which part of the child's result is collected. One name, not the row: what
    # crosses the boundary is a list of values for an IN comparison, so the child's
    # other columns are never read by the parent and never leave the database.
    #
    # For a tool-config child it is a column of its result. For a graph child it is a
    # key of the graph's last output — the same name, because a graph node's rows are
    # dictionaries exactly as a query's are, and the one case where a graph returns a
    # bare list rather than rows is handled by reading the list itself.
    child_column: Mapped[str] = mapped_column(String(255), nullable=False)

    # Where those values land in the parent, and what it means depends on the
    # parent's query_mode — the same split the parent's own query already has:
    #
    #   builder → a column reference ("client_id" or "projects.client_id"),
    #             validated against the tables the parent's query reads;
    #   sql     → the name of a bind parameter, written as `:name` in the
    #             statement. The statement is never rewritten; the list is bound
    #             at execution as an expanding parameter.
    #
    # One column rather than two nullable ones: a link always has exactly one
    # target, and which kind it is is already recorded on the parent.
    parent_reference: Mapped[str] = mapped_column(String(255), nullable=False)

    # Whether those values are matched all at once or iterated. See BINDING_MODES.
    # A parent may have at most one `each` child, refused when the link is saved:
    # two would be a cartesian product of two result sets, and the row cap makes
    # that a truncated answer rather than a bigger one.
    binding_mode: Mapped[str] = mapped_column(
        String(16),
        nullable=False,
        default=BINDING_MODE_IN_LIST,
        server_default=BINDING_MODE_IN_LIST,
    )

    # For an iterating link only: the name under which each row records the value
    # that produced it. Rows from twenty runs of one statement are otherwise
    # indistinguishable once concatenated, and a statement that filters on a
    # department without selecting it is perfectly ordinary SQL — so this is what
    # makes a fan-out result groupable by the thing it was fanned out over.
    #
    # Optional, because a query that already returns the value needs no second
    # copy of it. Asking for one anyway is refused at run time as a column
    # collision rather than overwriting what the database returned.
    value_alias: Mapped[str] = mapped_column(String(255), nullable=True)

    # Evaluation and display order among a parent's children. Meaningful: children
    # run in this order and the first one to return nothing stops the chain, so an
    # operator can put the cheapest or most selective child first.
    position: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
        default=0,
        server_default="0",
    )

    created_at: Mapped[DateTime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        index=True,
    )

    updated_at: Mapped[DateTime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        index=True,
    )

    __table_args__ = (
        # The same child may feed a parent twice — a tool returning client ids
        # could restrict both `owner_id` and `billed_to_id` — so the target is
        # part of the key. What is refused is the same child bound to the same
        # target twice, which would AND a list against itself.
        UniqueConstraint(
            "parent_id",
            "child_id",
            "parent_reference",
            name="uq_tool_config_links_parent_child_target",
        ),
        # The same rule for a graph child. A separate constraint rather than one
        # three-column one, because with `child_id` NULL Postgres treats every row as
        # distinct and graph links would never collide in the constraint above.
        UniqueConstraint(
            "parent_id",
            "child_graph_id",
            "parent_reference",
            name="uq_tool_config_links_parent_graph_target",
        ),
        # Exactly one child. A CHECK rather than a service rule — unlike the
        # attachment exclusivity on `tool_graphs`, which has an operator on the far
        # side who needs a sentence. No form can produce a link with two children or
        # none; that would be a bug in this application, and a constraint is how a bug
        # is stopped from becoming data.
        CheckConstraint(
            "(child_id IS NOT NULL) <> (child_graph_id IS NOT NULL)",
            name="ck_tool_config_links_one_child",
        ),
        # Reading a parent's children in the order they run is the hot path: it
        # happens on every nested tool call and on every list render.
        Index("ix_tool_config_links_parent_position", "parent_id", "position"),
    )

File: static/js/tool_chain.js
/**
 * Nested Tools — the repeating rows that say which tools this one embeds.
 *
 * One row is one sub-query: *run `child`, take its `column`, and restrict this
 * query at `target`*. What `target` means follows the query mode live — a column
 * of this query in builder mode, the name of a `:placeholder` in SQL mode — so
 * switching mode re-draws the rows rather than leaving a control that means
 * nothing.
 *
 * A row also says *how*: match any of the values (one run), or run this query once
 * per value. The second reveals a second line — the name each row records its value
 * under — because that is the only field that means anything for it, and a control
 * that is always visible and usually inert reads as a field the user forgot.
 *
 * Its own file rather than a third IIFE in tool_configs.js, which is already a
 * thousand lines of query builder. Nothing is shared between them: this card is
 * about *other tools*, the builder is about columns of one query, and they meet
 * only in the form that posts both.
 *
 * Conventions copied from tool_configs.js because they are load-bearing:
 * rows are built with createElement and never innerHTML (every name here comes
 * out of the user's own database), the hidden JSON field is the single output,
 * and init is idempotent so an htmx swap cannot wire the same card twice.
 *
 * Everything here is convenience. tool_chain_service re-validates every link on
 * save — same owner, same datasource, enabled, no cycle, within the depth and
 * size caps — so a hand-posted payload is checked by the same rules a click is.
 */
(function () {
    "use strict";

    var CARD_SELECTOR = "[data-nested-tools]";
    var OPTIONS_URL = "/tool-configs/child-options";

    document.addEventListener("DOMContentLoaded", function () {
        scan(document);
    });

    document.addEventListener("htmx:load", function (event) {
        scan(event.target);
    });

    scan(document);

    /** @param {Document|Element} root */
    function scan(root) {
        if (!root || !root.querySelectorAll) return;

        if (root.matches && root.matches(CARD_SELECTOR)) init(root);
        Array.prototype.forEach.call(root.querySelectorAll(CARD_SELECTOR), init);
    }

    /** @param {Element} root */
    function init(root) {
        if (root.dataset.nestedReady === "1") return;
        root.dataset.nestedReady = "1";

        var data = readData(root);
        var state = normalise(data.children);
        var rowsField = root.querySelector("[data-nested-rows]");
        var jsonField = root.querySelector("[data-nested-json]");
        var addButton = root.querySelector("[data-nested-add]");

        // Filled by the one fetch below. Until it answers, a saved row still
        // renders — from what it holds, not from what is offered — so an edit form
        // never briefly shows an empty picker over a real link.
        var options = [];

        if (addButton) {
            addButton.addEventListener("click", function () {
                state.push(blankRow());
                render();
            });
        }

        watchQueryMode(root, render);
        loadOptions();
        render();

        /** Fetch the tools that may be embedded, then re-render with real names. */
        function loadOptions() {
            if (!data.datasource_id) return;

            var url = OPTIONS_URL +
                "?datasource_id=" + encodeURIComponent(data.datasource_id) +
                (data.exclude ? "&exclude=" + encodeURIComponent(data.exclude) : "");

            request(url).then(function (payload) {
                options = (payload && payload.tools) || [];
                render();
            }).catch(function () {
                // Left to the saved rows: the picker is a convenience, and the
                // server refuses anything invalid anyway. A failed fetch must not
                // clear links the user already has.
                options = [];
            });
        }

        function render() {
            if (!rowsField) return;

            rowsField.textContent = "";

            if (!state.length) {
                rowsField.appendChild(mutedNote(
                    "No nested tools — this query runs on its own."
                ));
                sync();
                return;
            }

            state.forEach(function (entry, index) {
                rowsField.appendChild(buildRow(entry, index));
            });

            sync();
        }

        /**
         * One row: which tool, which of its columns, and where the values land.
         */
        function buildRow(entry, index) {
            var row = element("div", "mb-3");

            var top = element("div", "row g-2 align-items-center");

            top.appendChild(labelCell("Run"));
            top.appendChild(wrap("col-md-3", toolSelect(entry, index)));
            top.appendChild(labelCell("take"));
            top.appendChild(wrap("col-md-2", columnControl(entry)));
            top.appendChild(labelCell(isSqlMode() ? "fill" : "and filter"));
            top.appendChild(wrap("col-md-2", targetControl(entry)));
            top.appendChild(wrap("col-md-2", bindingModeControl(entry, index)));
            top.appendChild(wrap("col-md-auto", removeButton(function () {
                state.splice(index, 1);
                render();
            })));

            row.appendChild(top);

            if (entry.binding_mode === "each") {
                row.appendChild(aliasLine(entry));
            }

            return row;
        }

        /**
         * How the values are used: matched all at once, or one run each.
         *
         * Changing it re-renders rather than only syncing, because the alias line
         * below appears and disappears with it.
         */
        function bindingModeControl(entry, index) {
            return pairSelect(
                [
                    ["in_list", "match any"],
                    ["each", "run once per value"],
                ],
                entry.binding_mode || "in_list",
                function (value) {
                    state[index].binding_mode = value;
                    // A list binding has no single value to record, and the server
                    // refuses an alias on one — so clearing it here is the same rule
                    // said earlier rather than a second one.
                    if (value !== "each") state[index].value_alias = "";
                    render();
                }
            );
        }

        /** The name each row records the value it was produced for under. */
        function aliasLine(entry) {
            var line = element("div", "row g-2 align-items-center mt-1");

            line.appendChild(labelCell("record the value as"));
            line.appendChild(wrap("col-md-3", textInput(
                entry.value_alias,
                "column name (optional)",
                function (value) {
                    entry.value_alias = value;
                    sync();
                }
            )));
            line.appendChild(hint(
                "Leave blank if the query already returns the value."
            ));

            return line;
        }

        /**
         * The tool or graph to embed. A saved choice survives a picker that cannot
         * offer it.
         *
         * One select for both kinds, because the operator is making one choice: what
         * runs first and supplies the values. A graph is labelled as one, since it is
         * the only entry that may stop and ask somebody a question mid-run.
         */
        function toolSelect(entry, index) {
            var choices = options.map(function (tool) {
                var label = tool.kind === "graph"
                    ? tool.tool_name + " (graph)"
                    : tool.tool_name;

                return [tool.uuid, label];
            });

            if (entry.child_id && !options.some(function (tool) {
                return tool.uuid === entry.child_id;
            })) {
                choices.unshift([entry.child_id, entry.child_name || "(saved)"]);
            }

            // A new row shows the first choice, so take it as chosen. Written
            // straight onto the entry rather than through the change handler:
            // that handler re-renders, and a re-render from inside the render
            // loop appends every row twice.
            if (!entry.child_id && choices.length) {
                entry.child_id = choices[0][0];
                entry.child_kind = kindOf(choices[0][0]);
            }

            var select = pairSelect(choices, entry.child_id, function (value) {
                state[index].child_id = value;
                // Which key this row posts under follows the chosen thing, so it is
                // read here rather than inferred at save time from a list that may
                // have been reloaded since.
                state[index].child_kind = kindOf(value);
                // A different child returns different columns, so the column beside
                // it stops meaning anything.
                state[index].child_column = "";
                render();
            });

            if (!choices.length) {
                select.appendChild(option("", "Nothing available to embed"));
                select.disabled = true;
            }

            return select;
        }

        /** Whether a uuid names a graph or a tool config, per the loaded picker. */
        function kindOf(uuid) {
            var found = options.filter(function (candidate) {
                return candidate.uuid === uuid;
            })[0];

            return (found && found.kind) === "graph" ? "graph" : "tool";
        }

        /**
         * Which column of the child's result is collected.
         *
         * A dropdown when the child's output is knowable, a text input when it is
         * not — a SQL-mode tool, or a builder tool that selects everything. The
         * chain checks a typed name against the real result when it runs, which is
         * the same bargain the routing prompt makes about a SQL tool's fields.
         */
        function columnControl(entry) {
            var tool = options.filter(function (candidate) {
                return candidate.uuid === entry.child_id;
            })[0];

            if (tool && tool.columns && tool.columns.length) {
                if (!entry.child_column) entry.child_column = tool.columns[0];

                return valueSelect(tool.columns, entry.child_column, function (value) {
                    entry.child_column = value;
                    sync();
                });
            }

            return textInput(entry.child_column, "column name", function (value) {
                entry.child_column = value;
                sync();
            });
        }

        /**
         * Where the values land in this query.
         *
         * Builder mode offers the columns of the tables this tool reads; SQL mode
         * takes the name of a placeholder, which has to be in the statement — the
         * server refuses the save otherwise rather than letting it fail at run time.
         */
        function targetControl(entry) {
            if (isSqlMode()) {
                return textInput(
                    entry.parent_reference,
                    ":name used in the SQL",
                    function (value) {
                        entry.parent_reference = value;
                        sync();
                    }
                );
            }

            var columns = parentColumns();

            if (!columns.length) {
                return textInput(entry.parent_reference, "column", function (value) {
                    entry.parent_reference = value;
                    sync();
                });
            }

            if (!entry.parent_reference) entry.parent_reference = columns[0];

            return valueSelect(columns, entry.parent_reference, function (value) {
                entry.parent_reference = value;
                sync();
            });
        }

        /**
         * The columns this query can be filtered on: the base table's bare names,
         * plus every selected table's qualified ones.
         *
         * Mirrors what `query_joins.validated_column_reference` accepts — a bare
         * name means the base table, a qualified one must be a table the query
         * reads. A column of a table that is selected but not joined is offered and
         * refused on save, with a message naming it; guessing the live join chain
         * from here would mean re-implementing the builder's state.
         */
        function parentColumns() {
            var names = [];

            (data.column_map[data.base_table] || []).forEach(function (column) {
                names.push(column);
            });

            Object.keys(data.column_map).forEach(function (table) {
                if (table === data.base_table) return;
                (data.column_map[table] || []).forEach(function (column) {
                    names.push(table + "." + column);
                });
            });

            return names;
        }

        /** Whether the form currently means to save a SQL statement. */
        function isSqlMode() {
            var form = root.closest("form");
            var option = form && form.querySelector('[data-query-mode-option="sql"]');

            return Boolean(option && option.checked);
        }

        /**
         * The rows, as the server reads them.
         *
         * A graph posts its uuid as `child_graph_id` and a tool config as `child_id`,
         * which is what `tool_config_links` stores — exactly one of the two per row. The
         * browser keeps a single `child_id` field for either, because one select writes
         * it; the split happens here, once, where the payload is built.
         */
        function sync() {
            if (!jsonField) return;

            var rows = state.filter(function (entry) {
                return entry.child_id;
            }).map(function (entry) {
                var row = {
                    child_column: entry.child_column,
                    parent_reference: entry.parent_reference,
                    binding_mode: entry.binding_mode,
                    value_alias: entry.value_alias
                };

                if (entry.child_kind === "graph") {
                    row.child_graph_id = entry.child_id;
                } else {
                    row.child_id = entry.child_id;
                }

                return row;
            });

            jsonField.value = JSON.stringify(rows);
        }
    }

    /**
     * Re-render when the query mode changes.
     *
     * The rows themselves change shape with the mode, so this is not cosmetic: a
     * placeholder name left in a column dropdown would be saved as a column.
     */
    function watchQueryMode(root, onChange) {
        var form = root.closest("form");
        if (!form) return;

        Array.prototype.forEach.call(
            form.querySelectorAll("[data-query-mode-option]"),
            function (option) {
                option.addEventListener("change", onChange);
            }
        );
    }

    // ----------------------------------------------------------------------
    // State
    // ----------------------------------------------------------------------

    function readData(root) {
        var block = root.querySelector("[data-nested-data]");
        var parsed = {};

        try {
            parsed = JSON.parse((block && block.textContent) || "{}");
        } catch (error) {
            parsed = {};
        }

        return {
            datasource_id: parsed.datasource_id || "",
            exclude: parsed.exclude || "",
            base_table: parsed.base_table || "",
            column_map: parsed.column_map || {},
            children: Array.isArray(parsed.children) ? parsed.children : [],
        };
    }

    /**
     * Saved rows as the editor holds them.
     *
     * Both child keys collapse into one `child_id`, with `child_kind` remembering which
     * it was. The select writes one field, so keeping two would mean every read having
     * to ask which one is set — and a row where both looked set would have no answer.
     */
    function normalise(children) {
        return children.map(function (entry) {
            var isGraph = Boolean(entry.child_graph_id)
                || entry.child_kind === "graph";

            return {
                child_id: text(entry.child_graph_id || entry.child_id),
                child_kind: isGraph ? "graph" : "tool",
                child_name: text(entry.child_name),
                child_column: text(entry.child_column),
                parent_reference: text(entry.parent_reference),
                binding_mode: text(entry.binding_mode) || "in_list",
                value_alias: text(entry.value_alias),
            };
        });
    }

    function blankRow() {
        return {
            child_id: "",
            child_kind: "tool",
            child_name: "",
            child_column: "",
            parent_reference: "",
            binding_mode: "in_list",
            value_alias: "",
        };
    }

    // ----------------------------------------------------------------------
    // Controls — createElement only; every name here is user data
    // ----------------------------------------------------------------------

    function valueSelect(values, selected, onChange) {
        return pairSelect(
            values.map(function (value) { return [value, value]; }),
            selected,
            onChange
        );
    }

    /**
     * A <select> over [value, label] pairs.
     *
     * A selected value the options do not contain is kept as an option of its own,
     * so a link to a tool the picker could not load still shows what it holds
     * instead of silently becoming the first choice.
     *
     * `onChange` fires on a real change only. Building a control never writes back
     * to the state it was built from — the caller settles the default first, so a
     * render cannot restart itself part-way through and append the rows twice.
     */
    function pairSelect(pairs, selected, onChange) {
        var select = element("select", "form-select form-select-sm");
        var choices = pairs.slice();

        if (selected && !choices.some(function (pair) { return pair[0] === selected; })) {
            choices.unshift([selected, selected]);
        }

        choices.forEach(function (pair) {
            select.appendChild(option(pair[0], pair[1], pair[0] === selected));
        });

        select.addEventListener("change", function () {
            onChange(select.value);
        });

        return select;
    }

    function textInput(value, placeholder, onInput) {
        var input = element("input", "form-control form-control-sm font-monospace");
        input.type = "text";
        input.value = value || "";
        input.placeholder = placeholder;
        input.addEventListener("input", function () {
            onInput(input.value.trim());
        });
        return input;
    }

    function removeButton(onClick) {
        var button = element("button", "btn btn-sm btn-outline-danger");
        button.type = "button";
        button.textContent = "×";
        button.title = "Remove this nested tool";
        button.addEventListener("click", onClick);
        return button;
    }

    function labelCell(label) {
        var cell = element("div", "col-md-auto text-muted small");
        cell.textContent = label;
        return cell;
    }

    function hint(message) {
        var cell = element("div", "col-md-auto text-muted small");
        cell.textContent = message;
        return cell;
    }

    function mutedNote(message) {
        var note = element("p", "text-muted small mb-0");
        note.textContent = message;
        return note;
    }

    function wrap(className, child) {
        var cell = element("div", className);
        cell.appendChild(child);
        return cell;
    }

    function option(value, label, selected) {
        var node = document.createElement("option");
        node.value = value;
        node.textContent = label;
        if (selected) node.selected = true;
        return node;
    }

    function element(tag, className) {
        var node = document.createElement(tag);
        if (className) node.className = className;
        return node;
    }

    function text(value) {
        return value === null || value === undefined ? "" : String(value);
    }

    /** JSON GET, through the app's fetch wrapper when one is present. */
    function request(url) {
        if (typeof window.safeFetch === "function") {
            return window.safeFetch(url).then(function (response) {
                return response.json();
            });
        }

        return fetch(url, { headers: { Accept: "application/json" } })
            .then(function (response) { return response.json(); });
    }
})();
